bigtable/client/polling_policy.h

```diff
@@ -0,0 +1,104 @@
+// Copyright 2018 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_POLLING_POLICY_H_
+#define GOOGLE_CLOUD_CPP_BIGTABLE_CLIENT_POLLING_POLICY_H_
+
+#include "bigtable/client/rpc_backoff_policy.h"
+#include "bigtable/client/rpc_retry_policy.h"
+#include "bigtable/client/version.h"
+#include <grpc++/grpc++.h>
+
+namespace bigtable {
+inline namespace BIGTABLE_CLIENT_NS {
+/**
+ * Define the interface for providing asynchronous repetitive call rules
+ *
+ */
+class PollingPolicy {
+ public:
+  virtual ~PollingPolicy() = default;
+
+  /**
+   * Return a new copy of this object.
+   *
+   * Typically implemented as
+   * @code
+   *   return std::unique_ptr<PollingPolicy>(new Foo(*this));
+   * @endcode
+   */
+  virtual std::unique_ptr<PollingPolicy> clone() = 0;
+
+  /**
+   * Return true if `status` represents a permanent error that cannot be
+   * retried.
+   */
+  virtual bool IsPermanentError(grpc::Status const& status) = 0;
+
+  /**
+   * Handle an RPC failure.
+   *
+   * @return true if the RPC operation should be retried.
+   */
+  virtual bool OnFailure(grpc::Status const& status) = 0;
+
+  /**
+   * Return true if we cannot try again.
+   */
+  virtual bool Exhausted() = 0;
+
+  /**
+   * Return for how long we should wait before trying again.
+   */
+  virtual std::chrono::milliseconds WaitPeriod() = 0;
+};
+
+template <typename Retry = LimitedTimeRetryPolicy,
+          typename Backoff = ExponentialBackoffPolicy>
+class GenericPollingPolicy : public PollingPolicy,
+                             private Retry,
+                             private Backoff {
+ public:
+  GenericPollingPolicy() : Retry(), Backoff() {}
+  GenericPollingPolicy(Retry retry, Backoff backoff)
+      : Retry(std::move(retry)), Backoff(std::move(backoff)) {}
+
+  using LimitedTimeRetryPolicy::clone;
```

Comment:
Maybe this is not ready for review yet, but I noticed this change.  This will not work (I think it will not compile), if the template is instantiated with another pair of policies.  Do you need these `using` statements?

Reply:
When this statement was not there, there was compilation error saying that multiple clone functions with different declaration(I don't remember exact error string).
Then I checked in stackoverflow for the solution and there it was mentioned to use the using with each parent clone function. I will again replicate same condition to see the error, but it seems it is compiled in travis and all checks are passed.
I also added OnFailure function as it was returning proper true and false for the retry conditions, without this function, test cases were not succeeded.

Reply:
Here is the error build
In file included from /v/bigtable/client/polling_policy.cc:15:
/v/bigtable/client/polling_policy.h:77:34: error: 
      'bigtable::v0::GenericPollingPolicy<bigtable::v0::LimitedTimeRetryPolicy,
      bigtable::v0::ExponentialBackoffPolicy>::clone' hides overloaded virtual
      functions [-Werror,-Woverloaded-virtual]
  std::unique_ptr<PollingPolicy> clone() override {
                                 ^
/v/bigtable/client/polling_policy.cc:20:45: note: in instantiation of template
      class
      'bigtable::v0::GenericPollingPolicy<bigtable::v0::LimitedTimeRetryPolicy,
      bigtable::v0::ExponentialBackoffPolicy>' requested here
  return std::unique_ptr<PollingPolicy>(new GenericPollingPolicy<>());

https://travis-ci.org/GoogleCloudPlatform/google-cloud-cpp/jobs/376158579

Reply:
if the template is instantiated with another pair of policies. Do you need these using statements?
You are right
I will check how to solve above problem without these using statements.

Reply:
You could switch from private inheritance to private member variables to express composition. Worth a shot at least.